### PR TITLE
1700 Clean ascwds workplace - Convert format_date_fields

### PR DESCRIPTION
**Pull request by RoyPriceCS50**
## Description
Trello ticket [1700](https://trello.com/c/cm1nQBdV)

Converted format_date_fields to polars as cast_date_strings_to_dates.

New function lives in polars_utils > cleaning_utils because it is used in cleaning ascwds workplace, ingest ons and spss_csv_to_parquet.

## Testing
- [x] Unit tests passing
- [x] Successful [Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/statemachines/view/arn%3Aaws%3Astates%3Aeu-west-2%3A856699698263%3AstateMachine%3A1700-cln-wkp-5-Transform-ASCWDS-Data?type=standard)
- [ ] Outputs checked in Athena
- [ ] Outputs compared using [SageMaker script](https://eu-west-2.console.aws.amazon.com/sagemaker/home?region=eu-west-2#/notebook-instances/data-engineering-notebook)
      - Notebook name: compare_outputs.ipynb
- [ ] Output Screenshots added to Trello ticket

Outputs do not match pyspark version because other functions that filter workplace dataset are yet to be converted.

All the columns with "date" in name except import_date are date type.

## Migration to polars checklist
- [x] Check that similar utils functions don't already exist, or if one can be created
- [x] Check that utils are being saved in the appropriate place
- [x] Used LazyFrame option for test data
- [x] Unit tests added
- [x] Docstrings added
- [x] Added comment to pyspark functions which have been migrated with the link to the new location
      `# converted to polars -> filepath.py`
- [ ] Use new file structure when saving to S3
      `{dataset}_{sub_dataset_if_relevant}_{order_of_process_if_relevant}_{very_brief_description}`
- [x] Updated CHANGELOG
- [x] Code reviewed by AI
- [x] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py [hunk 1]
```diff
@@ -32,24 +32,16 @@ def main(
...
-    #     ascwds_workplace_df,
-    #     AWPClean.import_date,
-    #     AWPClean.ascwds_workplace_import_date,
-    # )
+    lf = cUtils.column_to_date(
+        lf, AWPClean.import_date, AWPClean.ascwds_workplace_import_date
+    )
+
+    # Temporarily casting import_date to string to prevent it being bigint.
+    lf = lf.with_columns(pl.col(AWPClean.import_date).cast(pl.String))
```
Comment (garypolzin) [inline projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py]: Pretty sure we don't use `import_date` after we've created `ascwds_workplace_import_date` so I'd drop the column, that's what we do elsewhere

```
    cqc_lf = column_to_date(
        cqc_lf, CQCLClean.import_date, CQCLClean.cqc_location_import_date
    ).drop(CQCLClean.import_date)
```

File: tests/test_polars_utils/test_cleaning_utils.py [hunk 2]
```diff
@@ -226,3 +227,64 @@ def test_create_banded_bed_count_column(self):
         )
 
         pl_testing.assert_frame_equal(returned_lf, expected_lf)
+
+
+class CastDateStringsToDatesTests(unittest.TestCase):
+    def test_casts_cols_that_match_arg_format_except_import_date(self):
+        test_lf_ddmmyyyy = pl.LazyFrame(
```
Comment (garypolzin) [inline tests/test_polars_utils/test_cleaning_utils.py]: remove all instances of `_ddmmyyyy` in all tests

File: polars_utils/cleaning_utils.py [hunk 3]
```diff
@@ -160,3 +160,37 @@ def create_banded_bed_count_column(
     return input_lf.with_columns(
         pl.when(is_not_care_home()).then(zero).otherwise(expr).alias(new_col)
     )
+
+
+def cast_date_strings_to_dates(
+    lf: pl.LazyFrame,
+    date_column_identifier: str = "date",
+    raw_date_format: str = None,
```
Comment (garypolzin) [inline polars_utils/cleaning_utils.py]: I'd set `raw_date_format: str = "%d/%m/%Y"` here too to avoid setting it ourselves each time - this function is only used twice in our main pipeline and it's in that format both times

(I'm ignoring the use in spss_csv_to_parquet.py as that hasn't been touched for 3 years and the 'snapshot date' column it uses in that job is no longer a concept, so that's unlikely to ever be updated)

File: tests/test_polars_utils/test_cleaning_utils.py [hunk 4]
```diff
@@ -226,3 +227,64 @@ def test_create_banded_bed_count_column(self):
...
+            data=[("01/01/2026", "02/01/2026", "words", None)],
+            schema={
+                "import_date": pl.String,
+                "any_other_date_col": pl.String,
+                "another_column": pl.String,
+                "null_date": pl.String,
+            },
+            orient="row",
```
Comment (garypolzin) [inline tests/test_polars_utils/test_cleaning_utils.py]: I'd probably avoid the last column being entirely null here and instead have two populated date columns (plus import date)

File: tests/test_polars_utils/test_cleaning_utils.py [hunk 5]
```diff
@@ -226,3 +227,64 @@ def test_create_banded_bed_count_column(self):
...
+        expected_lf_ddmmyyyy = pl.LazyFrame(
+            data=[(None, None)],
+            schema={
+                "time_element_date": pl.Date,
+                "malformed_date": pl.Date,
+            },
+            orient="row",
+        )
+
+        pl_testing.assert_frame_equal(returned_lf_ddmmyyyy, expected_lf_ddmmyyyy)
```
Comment (garypolzin) [inline tests/test_polars_utils/test_cleaning_utils.py]: potentially better to be one column with several date types, one of which is correct like

```
def test_invalid_date_strings_are_converted_to_null(self):
    input_lf = pl.LazyFrame(
        {
            "time_element_date": [
                "01/01/2026",        # valid
                "03/01/2026 00:00",  # invalid format
                "2026/01/03",        # wrong format
                "",                  # empty
                None,                # already null
            ]
        }
    )

    result_lf = job.cast_date_strings_to_dates(
        input_lf, raw_date_format="%d/%m/%Y"
    )

    expected_lf = pl.LazyFrame(
        {
            "time_element_date": [
                date(2026, 1, 1),
                None,
                None,
                None,
                None,
            ]
        }
    )

    pl_testing.assert_frame_equal(result_lf, expected_lf)
```

File: tests/test_polars_utils/test_cleaning_utils.py [hunk 6]
```diff
@@ -226,3 +227,64 @@ def test_create_banded_bed_count_column(self):
...
+
+    def test_when_raw_date_format_has_time_element(self):
+        test_lf_ddmmyyyy = pl.LazyFrame({"time_element_date": ["2026/01/01 00:00:00"]})
+        returned_lf_ddmmyyyy = job.cast_date_strings_to_dates(
+            test_lf_ddmmyyyy, raw_date_format="%Y/%m/%d %T"
+        )
+        expected_lf_ddmmyyyy = pl.LazyFrame({"time_element_date": date(2026, 1, 1)})
+
+        pl_testing.assert_frame_equal(returned_lf_ddmmyyyy, expected_lf_ddmmyyyy)
```
Comment (garypolzin) [inline tests/test_polars_utils/test_cleaning_utils.py]: this test is a bit redundant as it's not testing the function, it's just testing that polars knows how to parse a date time correctly

File: tests/test_polars_utils/test_cleaning_utils.py [hunk 7]
```diff
@@ -226,3 +227,64 @@ def test_create_banded_bed_count_column(self):
         )
 
         pl_testing.assert_frame_equal(returned_lf, expected_lf)
+
+
+class CastDateStringsToDatesTests(unittest.TestCase):
```
Comment (garypolzin) [inline tests/test_polars_utils/test_cleaning_utils.py:232]: covered below in more detail but I think two tests are fine here
- check that multiple date columns are converted but import_date isn't
- check that nulls are returned if the string format of the date doesn't match the required format

**Review by garypolzin — COMMENTED** [review]
Couple of minor tweaks and some edits to the tests

File: projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py (hunk 1 above)
Comment (RoyPriceCS50) [reply projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py]: I think we're going to need it for comparisons to pyspark version.

Comment (garypolzin) [reply projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py]: we can drop that col from the pyspark version when comparing the two

**Review by garypolzin — APPROVED** [review]
think it's best to remove import_date, pretty sure we can remove that column from the like for like check, but otherwise all looks good
